pydantic_ai_slim/pydantic_ai/_agent_graph.py

```diff
@@ -356,6 +356,12 @@ async def _make_request(
         message_history = await _process_message_history(
             ctx.state.message_history, ctx.deps.history_processors, build_run_context(ctx)
         )
+
+        if ctx.deps.usage_limits and ctx.deps.usage_limits.pre_request_token_check_with_overhead:
+            token_count = await ctx.deps.model.count_tokens(message_history)
+
+            ctx.deps.usage_limits.check_tokens(_usage.Usage(request_tokens=token_count.total_tokens))
```

Comment:
I think we should use `check_before_request` instead of `check_tokens`.

Also, I think this should be in `prepare_request` where we currently call `check_before_request`. I think likely move the `model_request_parameters` and `message_history` stuff there as well, and reduce the duplication between this method and `_stream`.